File: problems/堆/295.数据流的中位数.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class MedianFinder {
    priority_queue<int> small;
    priority_queue<int, vector<int>, greater<int>> large;
public:
    void addNum(int num) {
        small.push(num);
        large.push(small.top()); small.pop();
        if (small.size() < large.size()) {
            small.push(large.top()); large.pop();
        }
    }
    double findMedian() {
        return small.size() > large.size() ? small.top() : (small.top() + large.top()) / 2.0;
    }
};

int main() {
    // 测试用例
    MedianFinder mf;
    mf.addNum(1);
    mf.addNum(2);
    cout << mf.findMedian() << endl;  // 预期: 1.5
    mf.addNum(3);
    cout << mf.findMedian() << endl;  // 预期: 2.0
    MedianFinder mf2;
    mf2.addNum(2);
    mf2.addNum(3);
    mf2.addNum(4);
    cout << mf2.findMedian() << endl;  // 预期: 3.0
    return 0;
}

File: problems/堆/295.数据流的中位数.py
#!/usr/bin/env python3
"""
295. 数据流的中位数
"""
import heapq

class MedianFinder:
    def __init__(self):
        self.small = []
        self.large = []

    def add_num(self, num: int) -> None:
        heapq.heappush(self.small, -num)
        heapq.heappush(self.large, -heapq.heappop(self.small))
        if len(self.small) < len(self.large):
            heapq.heappush(self.small, -heapq.heappop(self.large))

    def find_median(self) -> float:
        if len(self.small) > len(self.large):
            return -self.small[0]
        return (-self.small[0] + self.large[0]) / 2.0

if __name__ == "__main__":
    # 测试用例
    mf = MedianFinder()
    mf.add_num(1)
    mf.add_num(2)
    print(mf.find_median())  # 预期: 1.5
    mf.add_num(3)
    print(mf.find_median())  # 预期: 2.0
    mf2 = MedianFinder()
    mf2.add_num(2)
    mf2.add_num(3)
    mf2.add_num(4)
    print(mf2.find_median())  # 预期: 3.0
